tidb-server/main.go

```diff
@@ -311,3 +272,221 @@ func flagBoolean(name string, defaultVal bool, usage string) *bool {
+	}
+
+	// Performance
+	if actualFlags[nmStatsLease] {
+		cfg.Performance.StatsLease = *statsLease
+	}
+	if actualFlags[nmJoinConcurrency] {
+		cfg.Performance.JoinConcurrency = *joinCon
+	}
+	if actualFlags[nmCrossJoin] {
+		cfg.Performance.CrossJoin = *crossJoin
+	}
+	if actualFlags[nmRetryLimit] {
+		cfg.Performance.RetryLimit = *retryLimit
+	}
+	if actualFlags[nmTCPKeepAlive] {
+		cfg.Performance.TCPKeepAlive = *tcpKeepAlive
+	}
+
+	// XProtocol
+	if actualFlags[nmStartXServer] {
+		cfg.XProtocol.XServer = *startXServer
+	}
+	if actualFlags[nmXHost] {
+		cfg.XProtocol.XHost = *xhost
+	}
+	if actualFlags[nmXPort] {
+		cfg.XProtocol.XPort, err = strconv.Atoi(*xport)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	if actualFlags[nmXSocket] {
+		cfg.XProtocol.XSocket = *xsocket
+	}
+}
+
+func validateConfig() {
+	if cfg.Security.SkipGrantTable && !hasRootPrivilege() {
+		log.Error("TiDB run with skip-grant-table need root privilege.")
+		os.Exit(-1)
+	}
+}
+
+func setGlobalVars() {
+	ddlLeaseDuration := parseLease(cfg.Lease)
+	tidb.SetSchemaLease(ddlLeaseDuration)
+	statsLeaseDuration := parseLease(cfg.Performance.StatsLease)
+	tidb.SetStatsLease(statsLeaseDuration)
+	ddl.RunWorker = cfg.RunDDL
+	tidb.SetCommitRetryLimit(cfg.Performance.RetryLimit)
+	plan.JoinConcurrency = cfg.Performance.JoinConcurrency
+	plan.AllowCartesianProduct = cfg.Performance.CrossJoin
+	privileges.SkipWithGrant = cfg.Security.SkipGrantTable
+}
+
+func setupLog() {
+	err := logutil.InitLogger(cfg.Log.ToLogConfig())
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func printInfo() {
+	// Make sure the TiDB info is always printed.
+	level := log.GetLevel()
+	log.SetLevel(log.InfoLevel)
+	printer.PrintTiDBInfo()
+	log.SetLevel(level)
+}
+
+func createServer() {
+	var driver server.IDriver
+	driver = server.NewTiDBDriver(storage)
+	var err error
+	svr, err = server.NewServer(cfg, driver)
+	if err != nil {
+		log.Fatal(errors.ErrorStack(err))
+	}
+	if cfg.XProtocol.XServer {
+		xcfg := &xserver.Config{
+			Addr:   fmt.Sprintf("%s:%d", cfg.XProtocol.XHost, cfg.XProtocol.XPort),
+			Socket: cfg.XProtocol.XSocket,
+		}
+		xsvr, err = xserver.NewServer(xcfg)
+		if err != nil {
+			log.Fatal(errors.ErrorStack(err))
+		}
+	}
+}
+
+func setupSignalHandler() {
+	sc := make(chan os.Signal, 1)
+	signal.Notify(sc,
+		syscall.SIGHUP,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+
+	go func() {
+		sig := <-sc
+		log.Infof("Got signal [%d] to exit.", sig)
+		if xsvr != nil {
+			xsvr.Close() // Should close xserver before server.
+		}
+		svr.Close()
+	}()
+}
+
+func setupMetrics() {
+	prometheus.MustRegister(timeJumpBackCounter)
+	go systimemon.StartMonitor(time.Now, func() {
+		timeJumpBackCounter.Inc()
+	})
+
+	pushMetric(cfg.Status.MetricsAddr, time.Duration(cfg.Status.MetricsInterval)*time.Second)
+}
+
+func runServer() {
+	if err := svr.Run(); err != nil {
+		log.Error(err)
+	}
+	if cfg.XProtocol.XServer {
+		if err := xsvr.Run(); err != nil {
```

Comment:
`srv.Run()` will not return, so it won't get to `xsvr.Run()` normally...

Reply:
Yes, but it's the same as before, we not going to handle this issue in this PR.

Reply:
okay @wentaoxu

Reply:
thx, @jackysp have fix it, will be merged with branch xwt/x-protocol